data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumerFactory.java

```diff
@@ -0,0 +1,98 @@
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.plugins.source.dynamodb.converter.StreamRecordConverter;
+import org.opensearch.dataprepper.plugins.source.dynamodb.coordination.EnhancedSourceCoordinator;
+import org.opensearch.dataprepper.plugins.source.dynamodb.coordination.partition.GlobalState;
+import org.opensearch.dataprepper.plugins.source.dynamodb.coordination.partition.StreamPartition;
+import org.opensearch.dataprepper.plugins.source.dynamodb.coordination.state.StreamProgressState;
+import org.opensearch.dataprepper.plugins.source.dynamodb.model.TableInfo;
+import org.opensearch.dataprepper.plugins.source.dynamodb.model.TableMetadata;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.dynamodb.streams.DynamoDbStreamsClient;
+
+import java.time.Instant;
+import java.util.Optional;
+
+/**
+ * Factory class to create shard consumers
+ */
+public class ShardConsumerFactory {
+    private static final Logger LOG = LoggerFactory.getLogger(ShardManager.class);
+
+    private final DynamoDbStreamsClient streamsClient;
+
+    private final EnhancedSourceCoordinator enhancedSourceCoordinator;
+    private final PluginMetrics pluginMetrics;
+    private final ShardManager shardManager;
+
+    private final Buffer<Record<Event>> buffer;
+
+    public ShardConsumerFactory(EnhancedSourceCoordinator enhancedSourceCoordinator, DynamoDbStreamsClient streamsClient, PluginMetrics pluginMetrics, ShardManager shardManager, Buffer<Record<Event>> buffer) {
+        this.streamsClient = streamsClient;
+        this.enhancedSourceCoordinator = enhancedSourceCoordinator;
+        this.pluginMetrics = pluginMetrics;
+        this.shardManager = shardManager;
+        this.buffer = buffer;
+
+    }
+
+    public Runnable createConsumer(StreamPartition streamPartition) {
+
+        LOG.debug("Try to create a thread for shard " + streamPartition.getShardId());
+
+        // Check and get the current state.
+        Optional<StreamProgressState> progressState = streamPartition.getProgressState();
+        String sequenceNumber = null;
+        Instant startTime = null;
+        boolean waitForExport = false;
+        if (progressState.isPresent()) {
+            sequenceNumber = progressState.get().getSequenceNumber();
+            waitForExport = progressState.get().shouldWaitForExport();
+            if (progressState.get().getStartTime() != 0) {
+                startTime = Instant.ofEpochMilli(progressState.get().getStartTime());
+            }
+        }
+
+        String shardIter = shardManager.getShardIterator(streamPartition.getStreamArn(), streamPartition.getShardId(), sequenceNumber);
+        if (shardIter == null) {
+            LOG.error("Unable to get a shard iterator, looks like the shard has expired");
```

Comment:
We should add something to this log to make it more useful. For example, the shardId or sequence number.